bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/events/OSGiEventManager.java

```diff
@@ -197,24 +204,24 @@ private Event createESHEvent(final EventFactory eventFactory, final String type,
 
     private void dispatchESHEvent(final Set<EventSubscriber> eventSubscribers, final Event event) {
         for (final EventSubscriber eventSubscriber : eventSubscribers) {
-            try {
-                EventFilter filter = eventSubscriber.getEventFilter();
-                if (filter == null || filter.apply(event)) {
-                    SafeMethodCaller.call(new ActionWithException<Void>() {
-
-                        @Override
-                        public Void call() throws Exception {
+            EventFilter filter = eventSubscriber.getEventFilter();
+            if (filter == null || filter.apply(event)) {
+                scheduler.submit(() -> {
+                    try {
+                        SafeMethodCaller.call((ActionWithException<Void>) () -> {
```

Comment:
This looks not efficient to me. You are submitting a job to be executed in a thread of the eventManager thread pool and then you do a (blocking) safe method call, which again spawns another thread.

Imho what is missing is a method in the SafeMethodCaller to do async (non-blocking) calls - if that existed, a single thread would be enough here.

Reply:
Will the order of events still be maintained when an eventSubscriber is receiving events asynchronously?

Reply:
Very good point - no, this was not guaranteed anymore. 
#4655 will take care of this by executing them _asynchronously_ but _sequentially_ (i.e. not in parallel).